standard/expressions.md

```diff
@@ -2370,6 +2371,77 @@ A *default_value_expression* is a constant expression ([§12.20](expressions.md#
 ### 12.7.16 Anonymous method expressions
 
 An *anonymous_method_expression* is one of two ways of defining an anonymous function. These are further described in [§12.16](expressions.md#1216-anonymous-function-expressions).
+	
+### 12.7(.x) §stack-allocation Stack allocation
+
+A stack allocation initializer allocates a block of memory from the call stack.
+
+```ANTLR
+stackalloc_initializer
+    : 'stackalloc' unmanaged_type '[' expression ']'
+    | 'stackalloc' unmanaged_type? '[' expression? ']' stackalloc_initializer_elements
+    ;
+
+stackalloc_initializer_elements
+    : '{' stackalloc_initializer_element_list? '}'
+    | '{' stackalloc_initializer_element_list ',' '}'
+    ;
+
+stackalloc_initializer_element_list
+    : stackalloc_element_initializer (',' stackalloc_element_initializer)*
+    ;
+    
+stackalloc_element_initializer
+    : expression
+    ;
+```
+
+The *unmanaged_type* ([§9.8](types.md#98-unmanaged-types)) indicates the type of the items that will be stored in the newly allocated location, and the *expression* indicates the number of these items. Taken together, these specify the required allocation size. As the size of a stack allocation cannot be negative, it is a compile-time error to specify the number of items as a *constant_expression* that evaluates to a negative value.
+
+If *unmanaged_type* is omitted, it is inferred from the corresponding *stackalloc_initializer_elements*. If *expression* is omitted from *stackalloc_initializer*, it is inferred to be the number of *stackalloc_element_initializer*s in the corresponding *stackalloc_initializer_elements*.
+
+When a *stackalloc_initializer* includes both *expression* and *stackalloc_initializer_elements* the number of elements in that *stackalloc_initializer_elements* shall match the value of *expression*.
+
+A stack allocation initializer of the form `stackalloc T[E]` requires `T` to be an *unmanaged_type* and `E` to be an expression implicitly convertible to type `int`. The operator allocates `E * sizeof(T)` bytes from the call stack, and the resulting type and value are determined, as follows:
+
+- If *unmanaged_type* is a *pointer_type* ([§23.3](unsafe-code.md#233-pointer-types)) or if *stackalloc_initializer* is used in a context that requires a *pointer_type*, the result is a pointer, of type `T*`, to the newly allocated block. If the context cannot be inferred, a pointer context is assumed. As pointer contexts require unsafe code, see [§23.3](unsafe-code.md#239-stack-allocation) for more information.
```

Comment:
It is not clear what "a context that requires a *pointer_type*" means here:

```C#
using System;

class Widget
{
    public static implicit operator Widget(Span<int> s) => null;
    public static unsafe implicit operator Widget(void* p) => null;
}

static class Class1
{
    static unsafe void S()
    {
        Widget w = stackalloc byte[1];
    }
}
```

The compiler chooses to return `byte*` from `stackalloc` and convert that to `void*`. However, the context does not seem to require a *pointer_type* as the context would accept `Span<int>`. Does the standard explain why the compiler does not choose `Span<byte>` and then report that it cannot be converted to `Span<int>`?